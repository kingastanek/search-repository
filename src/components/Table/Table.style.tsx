import styled from 'styled-components';

export const Wrapper = styled.div`
  width: 100%;
  overflow: scroll;
  margin-top: 1rem;
`;

export const NoDataTd = styled.td`
  box-shadow: 0px 3px 6px #00000029;
  padding: 0px 2rem;
  background-color: var(--theme-white);
  height: 75px;
  border-radius: 0.5rem;
`;

export const MainTable = styled.table`
  width: 100%;
  border-collapse: separate;
  border-spacing: 0 0.75rem;
  padding-bottom: 0.75rem;

  &:first-child td:first-child {
    border-top-left-radius: 0.5rem;
    border-color: var(--theme-white);
    border-bottom-left-radius: 0.5rem;
  }
  &:last-child td:last-child {
    border-top-right-radius: 0.5rem;
    border-color: var(--theme-white);
    border-bottom-right-radius: 0.5rem;
    background-color: var(--theme-white);
  }
`;

export const TableTh = styled.th`
  text-align: left;
  padding: 1rem;
  color: var(--theme-black);
  background-color: var(--theme-white);
  font-size: 0.875rem;
`;

export const TableTr = styled.tr`
  box-shadow: 0px 3px 6px #00000029;
  padding: 0px 1.5rem;
`;

export const TableTd = styled.td`
  padding: 1rem;
  line-height: 1;
  position: relative;
  background-color: var(--theme-white);
  font-size: 0.875rem;
`;

export const TableClearButton = styled.button`
  outline: none;
  background: transparent;
  border: 0;
  line-height: 1;
`;

export const SortableButtonContainer = styled.div`
  cursor: pointer;
  display: flex;
  flex-direction: column;
  position: relative;
`;

export const FlexWrapper = styled.div`
  display: flex;
  align-items: center;
`;
