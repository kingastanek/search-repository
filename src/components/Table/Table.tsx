import { useEffect, useState } from 'react';
import { Icon } from 'components';
import {
  TableTd,
  TableTr,
  TableTh,
  TableClearButton,
  NoDataTd,
  Wrapper,
  MainTable,
  SortableButtonContainer,
  FlexWrapper,
} from './Table.style';
import COLORS from 'styles/colors';

export type tTableHead = {
  sortable: boolean;
  name: string;
  key: string;
};

type tTableBody = {
  id: number;
  order: number;
};

type tTable = {
  tableHead: tTableHead[];
  tableBodyData: tTableBody[];
  setSortParams: (field: string, order: string) => void;
};

const Table = ({ tableHead, tableBodyData, setSortParams }: tTable) => {
  const [sort, setSort] = useState<{
    field: string;
    direction: string;
  }>({
    field: '',
    direction: 'asc',
  });

  useEffect(() => {
    if (!!sort?.field && setSortParams) {
      setSortParams(sort.field, sort.direction);
    }
  }, [setSortParams, sort]);

  const setSortable = (key: string) => {
    let direction = 'asc';

    if (sort?.field === key && sort?.direction === 'asc') {
      direction = 'desc';
    }

    setSort({ field: key, direction });
  };

  const renderSortIcons = (item: tTableHead) => {
    return (
      <TableClearButton onClick={() => setSortable(item.key)}>
        <SortableButtonContainer>
          <div style={{ marginBottom: '-0.5rem' }}>
            <Icon
              icon='arrowSortUp'
              fill={`${
                !!sort && sort.field === item.key && sort.direction === 'asc'
                  ? COLORS.BLACK
                  : COLORS.GREY
              }`}
            />
          </div>
          <div style={{ marginTop: '-0.5rem' }}>
            <Icon
              icon='arrowSortDown'
              fill={`${
                !!sort && sort.field === item.key && sort.direction === 'desc'
                  ? COLORS.BLACK
                  : COLORS.GREY
              }`}
            />
          </div>
        </SortableButtonContainer>
      </TableClearButton>
    );
  };

  return (
    <Wrapper>
      <MainTable>
        <thead>
          <tr>
            {tableHead.map((item: tTableHead) => (
              <TableTh key={item.key}>
                <FlexWrapper>
                  {item.name}
                  {item.sortable && renderSortIcons(item)}
                </FlexWrapper>
              </TableTh>
            ))}
          </tr>
        </thead>
        {tableBodyData?.length ? (
          <tbody>
            {tableBodyData.map((item: tTableBody) => (
              <TableTr key={item.id}>
                {tableHead.map((tableHeadItem: tTableHead) => {
                  return (
                    <TableTd key={tableHeadItem.key}>
                      {item[tableHeadItem.key]}
                    </TableTd>
                  );
                })}
              </TableTr>
            ))}
          </tbody>
        ) : (
          <tbody>
            <tr>
              <NoDataTd>No results</NoDataTd>
            </tr>
          </tbody>
        )}
      </MainTable>
    </Wrapper>
  );
};

export default Table;
