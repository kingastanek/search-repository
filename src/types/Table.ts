export type tTableHead = {
  sortable: boolean;
  name: string;
  key: string;
};

export type tTableBody = {
  id: number;
  order: number;
};

export type tSortParams = {
  fieldName: string;
  order: string;
};