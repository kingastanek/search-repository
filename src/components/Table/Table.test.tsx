import { shallow } from 'enzyme';
import Table from './Table';

describe('Table', () => {
  const basicTableHead = [
    {
      sortable: false,
      name: 'Lp.',
      key: 'order',
    },
    {
      sortable: false,
      name: 'Name',
      key: 'name',
    },
  ];

  const basicTableBody = [
    {
      id: 1,
      order: 1,
      name: 'Name 123',
    },
  ];

  const configuredTableHead = [
    {
      sortable: false,
      name: 'Lp.',
      key: 'order',
    },
    {
      sortable: false,
      name: 'Name',
      key: 'name',
    },
  ];

  const configuredTableBody = [
    {
      id: 1,
      order: 1,
      name: 'Name 1',
    },
    {
      id: 2,
      order: 2,
      name: 'Name 2',
    },
  ];

  it('should render basic table - no sort, no marked items, no actions', () => {
    const component = shallow(
      <Table
        tableHead={basicTableHead}
        tableBodyData={basicTableBody}
        setSortParams={() => jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render table with configuration - sort, marked items, actions', () => {
    const component = shallow(
      <Table
        tableHead={configuredTableHead}
        tableBodyData={configuredTableBody}
        setSortParams={() => jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('should render table with no results info', () => {
    const component = shallow(
      <Table
        tableHead={configuredTableHead}
        tableBodyData={[]}
        setSortParams={() => jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
