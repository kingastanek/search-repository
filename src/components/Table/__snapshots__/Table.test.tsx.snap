// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render basic table - no sort, no marked items, no actions 1`] = `ShallowWrapper {}`;

exports[`Table should render table with configuration - sort, marked items, actions 1`] = `ShallowWrapper {}`;

exports[`Table should render table with no results info 1`] = `ShallowWrapper {}`;
